src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Refactoring/ExtractToCodeBehindCodeActionResolver.cs

```diff
@@ -0,0 +1,156 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.AspNetCore.Razor.LanguageServer.ProjectSystem;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Razor;
+using Newtonsoft.Json.Linq;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+
+using CSharpSyntaxFactory = Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Refactoring
+{
+    class ExtractToCodeBehindCodeActionResolver : RazorCodeActionResolver
+    {
+        private readonly ForegroundDispatcher _foregroundDispatcher;
+        private readonly DocumentResolver _documentResolver;
+
+        public override string Action => "ExtractToCodeBehind";
+
+        public ExtractToCodeBehindCodeActionResolver(
+            ForegroundDispatcher foregroundDispatcher,
+            DocumentResolver documentResolver)
+        {
+            if (foregroundDispatcher is null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            if (documentResolver is null)
+            {
+                throw new ArgumentNullException(nameof(documentResolver));
+            }
+
+            _foregroundDispatcher = foregroundDispatcher;
+            _documentResolver = documentResolver;
+        }
+
+        override public async Task<WorkspaceEdit> ResolveAsync(JObject data, CancellationToken cancellationToken)
+        {
+            var actionParams = data.ToObject<ExtractToCodeBehindParams>();
+            var path = actionParams.Uri.LocalPath;
+
+            var document = await Task.Factory.StartNew(() =>
```

Comment:
Most resolvers will require a document snapshot so I'd recommend baking this document resolution into the `ResolveAsync` signature because you'll want to resolve the document once.

Reply:
Only one resolver is ever invoked at once, and since we know that the next provider/resolver won't need to access any document snapshots (CreateComponentFromTag), I'm not sure if this is the right course of action.

Reply:
> Only one resolver is ever invoked at once, and since we know that the next provider/resolver won't need to access any document snapshots (CreateComponentFromTag), I'm not sure if this is the right course of action.

Ah fair point. This isn't the "provide" portion 😄. I take back what I said. Feel free to leave it as is then